File: prototype/php/src/app/Actions/Notifications/NotifyTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Notifications;

use App\Domain\Money\Money;
use App\Domain\Notifications\NotificationMessage;
use App\Domain\Notifications\RecipientType;
use App\Models\Notification;

it('writes a row addressed to a seller', function (): void {
    $seller = $this->seller();

    $notification = app(Notify::class)(
        RecipientType::Seller,
        $seller->id,
        NotificationMessage::itemSold(4, Money::fromCents(9000)),
    );

    expect($notification)
        ->seller_id->toBe($seller->id)
        ->customer_id->toBeNull()
        ->subject->toBe('Item sold')
        ->read_at->toBeNull();
});

it('writes a row addressed to a customer', function (): void {
    $customer = $this->verifiedCustomer();

    $notification = app(Notify::class)(
        RecipientType::Customer,
        $customer->id,
        NotificationMessage::orderShipped(4, 'USPS', '9400111899')->withUrl('/orders/4'),
    );

    expect($notification)
        ->customer_id->toBe($customer->id)
        ->seller_id->toBeNull()
        ->url->toBe('/orders/4');
});

it('shows an unread notification for its recipient only', function (): void {
    $seller = $this->seller();
    $customer = $this->verifiedCustomer();
    $notify = app(Notify::class);

    $notify(RecipientType::Seller, $seller->id, NotificationMessage::itemSold(4, Money::fromCents(9000)));
    $notify(RecipientType::Customer, $customer->id, NotificationMessage::orderShipped(4, 'USPS', '94001'));

    expect(Notification::query()->unread()->for(RecipientType::Seller, $seller->id)->count())->toBe(1)
        ->and(Notification::query()->unread()->for(RecipientType::Customer, $customer->id)->count())->toBe(1);
});

File: prototype/rails/src/app/actions/notifications/notify_test.rb
require "commerce_test_case"

module Notifications
  class NotifyTest < CommerceTestCase
    def test_it_files_a_seller_message_under_the_seller
      shop = seller

      notification = Notify.new.call(
        recipient_type: Domain::Notifications::RecipientType::SELLER,
        recipient_id: shop.id,
        message: Domain::Notifications::NotificationMessage.item_sold(7, Domain::Money.from_cents(40_500))
      )

      assert_equal shop.id, notification.seller_id
      assert_nil notification.customer_id
      assert_equal "Item sold", notification.subject
    end

    def test_it_files_a_customer_message_under_the_customer
      buyer = customer

      notification = Notify.new.call(
        recipient_type: Domain::Notifications::RecipientType::CUSTOMER,
        recipient_id: buyer.id,
        message: Domain::Notifications::NotificationMessage.order_shipped(7, "USPS", "9400111899")
      )

      assert_equal buyer.id, notification.customer_id
      assert_nil notification.seller_id
    end

    def test_a_new_notification_is_unread
      notification = Notify.new.call(
        recipient_type: Domain::Notifications::RecipientType::SELLER,
        recipient_id: seller.id,
        message: Domain::Notifications::NotificationMessage.item_sold(7, Domain::Money.from_cents(40_500))
      )

      assert_includes Notification.unread, notification
    end

    def test_it_carries_the_url_the_message_points_at
      message = Domain::Notifications::NotificationMessage
                  .item_sold(7, Domain::Money.from_cents(40_500))
                  .with(url: "/seller/orders/7")

      notification = Notify.new.call(
        recipient_type: Domain::Notifications::RecipientType::SELLER,
        recipient_id: seller.id,
        message: message
      )

      assert_equal "/seller/orders/7", notification.url
    end
  end
end
